applications/ExaquteSandboxApplication/custom_processes/divergence_process.cpp

```diff
@@ -0,0 +1,180 @@
+            Vector values_y(number_nodes_element);
+            Vector values_z(number_nodes_element);
+            for(int i_node=0; i_node < static_cast<int>(number_nodes_element); ++i_node){
+                const auto &r_velocity = r_geometry[i_node].FastGetSolutionStepValue(VELOCITY);
+                values_x[i_node] = r_velocity[0];
+                values_y[i_node] = r_velocity[1];
+                if (dimension == 3) {
+                    values_z[i_node] = r_velocity[2];
+                }
+            }
+
+            // Set integration points
+            const auto& r_integration_method = r_geometry.GetDefaultIntegrationMethod(); // Default is 0
+            const auto& r_integration_points = r_geometry.IntegrationPoints(r_integration_method);
+            const std::size_t number_of_integration_points = r_integration_points.size(); // Default is 1
+
+            // Initialize auxiliary local variables
+            double aux_divergence = 0;
+            double aux_velocity_seminorm = 0;
+
+            // Get gradient shape functions and detJ0 on integration points
+            Vector detJ0;
+            r_geometry.ShapeFunctionsIntegrationPointsGradients(DN_DX, detJ0, r_integration_method);
+
+            // Loop over integration points
+            for ( IndexType point_number = 0; point_number < number_of_integration_points; ++point_number ){
+
+                // Compute local gradient
+                Vector grad_x;
+                Vector grad_y;
+                Vector grad_z;
+                grad_x = prod(trans(DN_DX[point_number]), values_x);
+                grad_y = prod(trans(DN_DX[point_number]), values_y);
+                if (dimension == 3) {
+                    grad_z = prod(trans(DN_DX[point_number]), values_z);
+                }
+
+                // Compute divergence and velocity seminorm
+                const double aux_current_divergence = ComputeAuxiliaryElementDivergence(grad_x, grad_y, grad_z);
+                const double aux_current_velocity_seminorm = ComputeAuxiliaryElementVelocitySeminorm(grad_x, grad_y, grad_z);
+                const double gauss_point_volume = r_integration_points[point_number].Weight() * detJ0[point_number];
+                aux_divergence += std::pow(aux_current_divergence,2) * gauss_point_volume;
+                aux_velocity_seminorm += aux_current_velocity_seminorm * gauss_point_volume;
+            }
+
+            // Set values to corresponding variables
+            it_elem->SetValue(DIVERGENCE,aux_divergence);
+            it_elem->SetValue(VELOCITY_H1_SEMINORM,aux_velocity_seminorm);
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    // Compute local auxiliar divergence
+    double DivergenceProcess::ComputeAuxiliaryElementDivergence(Vector& grad_x, Vector& grad_y, Vector& grad_z)
+    {
+        const std::size_t dimension = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+        double aux_current_divergence;
+        if (dimension == 2) {
```

Comment:
I think you can skip this if statement by passing an `std::vector<Vector>` and doing the corresponding summation in a loop.

Reply:
I removed the if statement by knowing the grad_z is zero for 2-dimensional problems.